torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
   return new_loop;
 }
 
+// Create a new vectorize For-Loop
+// Add For-Loop to If-Then-Else parent scope
+// for (index = start; index < extent; index += offset)
+// vectorize flag - Do not generate for-loop
+// shift value - Add shift to global indices generated within For-Loop
+void cloneVectorizeLoopNests(
+    kir::IfThenElse* parent_ite,
+    const std::vector<kir::ForLoop*>& for_loops,
+    kir::Val* extent,
+    bool vectorize,
+    kir::Val* shift) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set &&
+         first_expr->as<kir::UnaryOp>()->out()->isA<kir::TensorView>() &&
+         first_expr->as<kir::UnaryOp>()
+             ->out()
+             ->as<kir::TensorView>()
+             ->domain()
+             ->hasVectorize());
+
+    TORCH_INTERNAL_ASSERT(!has_vectorize_op || fl->body().exprs().size() == 1);
+
+    const auto new_loop = ir_builder.create<kir::ForLoop>(
+        fl->index(),
+        fl->iter_domain(),
+        vectorize && has_vectorize_op,
+        extent,
+        false,
+        shift);
+
+    for (auto expr : fl->body().exprs()) {
+      new_loop->body().push_back(expr);
+    }
+
+    parent_ite->thenBody().push_back(new_loop);
+  }
+}
+
+// Find any child For-Loops
+// Add remaining expressions to new parent For-Loop
+std::vector<kir::ForLoop*> parseVectorizedForLoop(
```

Comment:
What is this doing? Need more descriptive comments and some context.

Reply:
This function separates the allocations from the child for-loops. We need to duplicate the child-for loops. I see no reason to duplicate the allocations, so I move them to the top of the for-loop.